FirebaseSessions/Sources/Settings.swift

```diff
@@ -0,0 +1,167 @@
+//
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+
+protocol SettingsProtocol {
+  func loadCache(googleAppID: String, currentTime: Date)
+  var sessionsEnabled: Bool { get }
+  var samplingRate: Double { get }
+  var sessionTimeout: TimeInterval { get }
+}
+
+class Settings: SettingsProtocol {
+  private static let cacheDurationSecondsDefault: TimeInterval = 60 * 60
+  private let fileManager: SettingsFileManager
+  private let appInfo: ApplicationInfoProtocol
+  private let lock = NSLock()
+  private var _settingsDictionary: [String: AnyObject]?
+  private var isCacheKeyExpired: Bool
+  struct CacheKey: Codable {
+    var createdAt: Date
+    var googleAppID: String
+    var appVersion: String
+  }
+
+  var settingsDictionary: [String: AnyObject]? {
+    objc_sync_enter(lock)
+    defer { objc_sync_exit(lock) }
+    return _settingsDictionary
+  }
+
+  var sessionsEnabled: Bool {
+    guard let enabled = settingsDictionary?["sessions_enabled"]?.boolValue else {
+      return true
+    }
+    return enabled
+  }
+
+  var samplingRate: Double {
+    guard let rate = settingsDictionary?["sampling_rate"]?.doubleValue else {
+      return 1
+    }
+    return rate
+  }
+
+  var sessionTimeout: TimeInterval {
+    guard let timeout = settingsDictionary?["session_timeout"]?.doubleValue else {
+      return 30 * 60
+    }
+    return timeout
+  }
+
+  var isCacheExpired: Bool {
+    guard settingsDictionary != nil else {
+      return true
+    }
+    return isCacheKeyExpired
+  }
+
+  private var cacheDurationSeconds: TimeInterval {
+    guard let duration = settingsDictionary?["cache_duration"]?.doubleValue else {
+      return Settings.cacheDurationSecondsDefault
+    }
+    return duration
+  }
+
+  init(fileManager: SettingsFileManager = SettingsFileManager(), appInfo: ApplicationInfoProtocol) {
+    self.fileManager = fileManager
+    isCacheKeyExpired = false
+    self.appInfo = appInfo
+  }
+
+  func loadCache(googleAppID: String, currentTime: Date) {
+    guard let cacheData = fileManager.data(contentsOf: fileManager.settingsCacheContentPath) else {
+      Logger.logDebug("[Sessions:Settings] No settings were cached")
+      return
+    }
+    guard let parsedDictionary = cacheData.dictionaryValue else {
+      removeCache()
+      return
+    }
+    do {
+      objc_sync_enter(lock)
+      defer { objc_sync_exit(lock) }
```

Comment:
This is smart - if I understand correctly you're using `do {...}` to create a block so that the `defer {objc_sync_exit...}` can evaluate at the end of the `do`, instead of at the end of the function.

I think you may be able to improve readability by making a lock function with a callback.

Eg.

```
func locked(callback: () -> ()) {
      objc_sync_enter(lock)
      defer { objc_sync_exit(lock) }
      callback()
}
```

Then use it with
```
locked() {
  /// do threading stuff
}
```

Reply:
That's a great idea, I love this syntactic sugar. Updated the PR to do basically this, but using an extension so the syntax now look like:
```
lock.synchronized {
...
}
```